File: functions/invitations/accept_invitation.py
from datetime import datetime, timezone
from firebase_admin import firestore
from flask import abort
from models.constants import (
    Collections,
    InvitationFields,
    Status,
    FriendshipFields,
    ProfileFields,
)
from models.data_models import Friend
from utils.logging_utils import get_logger


def accept_invitation(request, invitation_id) -> Friend:
    """
    Accepts an invitation and creates a friendship between the users.

    This function:
    1. Checks if the invitation exists and is still valid
    2. Creates a new friendship document between the accepting user and the sender
    3. Deletes the invitation document

    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)
        invitation_id: The ID of the invitation to accept

    Returns:
        A Friend object representing the new friendship

    Raises:
        400: Invitation cannot be accepted (status: {status})
        400: Invitation has expired
        400: You cannot accept your own invitation
        404: Invitation not found
        404: User profile not found
        404: Sender profile not found
    """
    logger = get_logger(__name__)
    logger.info(f"User {request.user_id} accepting invitation {invitation_id}")

    current_user_id = request.user_id

    # Initialize Firestore client
    db = firestore.client()

    # Get the invitation document
    invitation_ref = db.collection(Collections.INVITATIONS).document(invitation_id)
    invitation_doc = invitation_ref.get()

    # Check if the invitation exists
    if not invitation_doc.exists:
        logger.warning(f"Invitation {invitation_id} not found")
        abort(404, description="Invitation not found")

    invitation_data = invitation_doc.to_dict()

    # Check invitation status
    status = invitation_data.get(InvitationFields.STATUS)
    if status != Status.PENDING:
        logger.warning(f"Invitation {invitation_id} has status {status}, not pending")
        abort(400, description=f"Invitation cannot be accepted (status: {status})")

    # Check if invitation has expired
    current_time = datetime.now(timezone.utc)
    expires_at = invitation_data.get(InvitationFields.EXPIRES_AT)
    if expires_at and isinstance(expires_at, datetime) and expires_at < current_time:
        # Update invitation status to expired
        invitation_ref.update({InvitationFields.STATUS: Status.EXPIRED})
        logger.warning(f"Invitation {invitation_id} has expired")
        abort(400, description="Invitation has expired")

    # Get the sender's user ID
    sender_id = invitation_data.get(InvitationFields.SENDER_ID)

    # Ensure the current user is not the sender (can't accept your own invitation)
    if sender_id == current_user_id:
        logger.warning(
            f"User {current_user_id} attempted to accept their own invitation {invitation_id}"
        )
        abort(400, description="You cannot accept your own invitation")

    # Get current user's profile
    current_user_profile_ref = db.collection(Collections.PROFILES).document(
        current_user_id
    )
    current_user_profile_doc = current_user_profile_ref.get()

    if not current_user_profile_doc.exists:
        logger.warning(f"Current user profile {current_user_id} not found")
        abort(404, description="User profile not found")

    current_user_profile = current_user_profile_doc.to_dict()

    # Get sender's profile
    sender_profile_ref = db.collection(Collections.PROFILES).document(sender_id)
    sender_profile_doc = sender_profile_ref.get()

    if not sender_profile_doc.exists:
        logger.warning(f"Sender profile {sender_id} not found")
        abort(404, description="Sender profile not found")

    sender_profile = sender_profile_doc.to_dict()

    # Create a batch operation for atomicity
    batch = db.batch()

    # Create a consistent friendship ID by sorting the user IDs
    user_ids = sorted([current_user_id, sender_id])
    friendship_id = f"{user_ids[0]}_{user_ids[1]}"

    # Check if friendship already exists
    friendship_ref = db.collection(Collections.FRIENDSHIPS).document(friendship_id)
    friendship_doc = friendship_ref.get()

    if friendship_doc.exists:
        friendship_data = friendship_doc.to_dict()
        status = friendship_data.get(FriendshipFields.STATUS)

        if status == Status.ACCEPTED:
            logger.warning(
                f"Users {current_user_id} and {sender_id} are already friends"
            )
            # Delete the invitation since they're already friends
            batch.delete(invitation_ref)
            batch.commit()

            # Return the existing friend using data from the friendship document
            if friendship_data.get(FriendshipFields.SENDER_ID) == sender_id:
                friend_name = friendship_data.get(FriendshipFields.SENDER_NAME, "")
                friend_username = friendship_data.get(
                    FriendshipFields.SENDER_USERNAME, ""
                )
                friend_avatar = friendship_data.get(FriendshipFields.SENDER_AVATAR, "")
            else:
                friend_name = friendship_data.get(FriendshipFields.RECEIVER_NAME, "")
                friend_username = friendship_data.get(
                    FriendshipFields.RECEIVER_USERNAME, ""
                )
                friend_avatar = friendship_data.get(
                    FriendshipFields.RECEIVER_AVATAR, ""
                )

            return Friend(
                user_id=sender_id,
                username=friend_username,
                name=friend_name,
                avatar=friend_avatar,
            )

    # Create the friendship document using profile data directly
    friendship_data = {
        FriendshipFields.SENDER_ID: sender_id,
        FriendshipFields.SENDER_NAME: sender_profile.get(ProfileFields.NAME, ""),
        FriendshipFields.SENDER_USERNAME: sender_profile.get(
            ProfileFields.USERNAME, ""
        ),
        FriendshipFields.SENDER_AVATAR: sender_profile.get(ProfileFields.AVATAR, ""),
        FriendshipFields.RECEIVER_ID: current_user_id,
        FriendshipFields.RECEIVER_NAME: current_user_profile.get(
            ProfileFields.NAME, ""
        ),
        FriendshipFields.RECEIVER_USERNAME: current_user_profile.get(
            ProfileFields.USERNAME, ""
        ),
        FriendshipFields.RECEIVER_AVATAR: current_user_profile.get(
            ProfileFields.AVATAR, ""
        ),
        FriendshipFields.STATUS: Status.ACCEPTED,
        FriendshipFields.CREATED_AT: current_time,
        FriendshipFields.UPDATED_AT: current_time,
        FriendshipFields.MEMBERS: [sender_id, current_user_id],
    }

    # Add operations to batch
    batch.set(friendship_ref, friendship_data)
    batch.delete(invitation_ref)

    # Commit the batch
    batch.commit()

    logger.info(
        f"User {current_user_id} accepted invitation {invitation_id} from {sender_id}"
    )

    # Return the friend object using sender's profile data
    return Friend(
        user_id=sender_id,
        username=sender_profile.get(ProfileFields.USERNAME, ""),
        name=sender_profile.get(ProfileFields.NAME, ""),
        avatar=sender_profile.get(ProfileFields.AVATAR, ""),
    )

File: functions/src/invitations/accept-invitation.ts
import { Request } from 'express';
import { getFirestore, Timestamp } from 'firebase-admin/firestore';
import { ApiResponse, EventName, InviteEventParams, } from '../models/analytics-events.js';
import {
  Collections,
  FriendPlaceholderTemplates,
  FriendshipFields,
  InvitationFields,
  ProfileFields,
  Status,
  UserSummaryFields,
} from '../models/constants.js';
import { Friend } from '../models/data-models.js';
import { BadRequestError, ForbiddenError } from '../utils/errors.js';
import { createFriendshipId, hasReachedCombinedLimit, } from '../utils/friendship-utils.js';
import {
  canActOnInvitation,
  getInvitationDoc,
  hasInvitationPermission,
  isInvitationExpired,
  updateInvitationStatus,
} from '../utils/invitation-utils.js';
import { getLogger } from '../utils/logging-utils.js';
import { createSummaryId, getProfileDoc, hasLimitOverride, } from '../utils/profile-utils.js';

import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Accepts an invitation and creates a friendship between the users.
 *
 * This function:
 * 1. Checks if the invitation exists and is still valid
 * 2. Creates a new friendship document between the accepting user and the sender
 * 3. Deletes the invitation document
 *
 * Validates that:
 * 1. The accepting user hasn't reached the combined limit of friends and active invitations (5)
 * 2. The sender hasn't reached the combined limit of friends and active invitations (5)
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - params: Route parameters containing:
 *                - invitation_id: The ID of the invitation to accept
 *
 * @returns An ApiResponse containing the new friend and analytics
 *
 * @throws 400: Invitation cannot be accepted (status: {status})
 * @throws 400: Invitation has expired
 * @throws 400: You cannot accept your own invitation
 * @throws 400: User has reached the maximum number of friends and active invitations
 * @throws 400: Sender has reached the maximum number of friends and active invitations
 * @throws 404: Invitation not found
 * @throws 404: User profile not found
 * @throws 404: Sender profile not found
 */
export const acceptInvitation = async (
  req: Request,
): Promise<ApiResponse<Friend>> => {
  const currentUserId = req.userId;
  const invitationId = req.params.invitation_id;

  logger.info(`User ${currentUserId} accepting invitation ${invitationId}`);

  // Initialize Firestore client
  const db = getFirestore();

  if (!invitationId) {
    throw new BadRequestError('Invitation ID is required');
  }

  // Get the invitation document
  const { ref: invitationRef, data: invitationData } =
    await getInvitationDoc(invitationId);

  // Check invitation status
  const status = invitationData[InvitationFields.STATUS];
  canActOnInvitation(status, 'accept');

  // Check if invitation has expired
  const expiresAt = invitationData[InvitationFields.EXPIRES_AT] as Timestamp;
  if (isInvitationExpired(expiresAt)) {
    // Update invitation status to expired
    await updateInvitationStatus(invitationRef, Status.EXPIRED);
    throw new ForbiddenError('Invitation has expired');
  }

  // Get the sender's user ID
  const senderId = invitationData[InvitationFields.SENDER_ID];

  // Ensure the current user is not the sender (can't accept your own invitation)
  hasInvitationPermission(senderId, currentUserId, 'accept');

  // Check combined limit for the accepting user
  const {
    friendCount: currentUserFriendCount,
    activeInvitationCount: currentUserInvitationCount,
    hasReachedLimit,
  } = await hasReachedCombinedLimit(currentUserId);
  if (hasReachedLimit) {
    const override = await hasLimitOverride(currentUserId);
    if (!override) {
      throw new BadRequestError(
        'You have reached the maximum number of friends and active invitations',
      );
    }
  }

  // Check combined limit for the sender (excluding this invitation)
  const { hasReachedLimit: senderHasReachedLimit } =
    await hasReachedCombinedLimit(senderId, invitationId);
  if (senderHasReachedLimit) {
    const override = await hasLimitOverride(senderId);
    if (!override) {
      throw new BadRequestError(
        'Sender has reached the maximum number of friends and active invitations',
      );
    }
  }

  // Get current user's profile
  const { data: currentUserProfile } = await getProfileDoc(currentUserId);

  // Get sender's profile
  const { data: senderProfile } = await getProfileDoc(senderId);

  // Create a batch operation for atomicity
  const batch = db.batch();

  // Create a consistent friendship ID by sorting the user IDs
  const friendshipId = createFriendshipId(currentUserId, senderId);

  // Check if friendship already exists
  const friendshipRef = db
    .collection(Collections.FRIENDSHIPS)
    .doc(friendshipId);
  const friendshipDoc = await friendshipRef.get();

  if (friendshipDoc.exists) {
    const friendshipData = friendshipDoc.data() || {};
    const friendshipStatus = friendshipData[FriendshipFields.STATUS];

    if (friendshipStatus === Status.ACCEPTED) {
      logger.warn(`Users ${currentUserId} and ${senderId} are already friends`);
      // Delete the invitation since they're already friends
      batch.delete(invitationRef);
      await batch.commit();

      // Return the existing friend using data from the friendship document
      let friendName: string;
      let friendUsername: string;
      let friendAvatar: string;

      if (friendshipData[FriendshipFields.SENDER_ID] === senderId) {
        friendName = friendshipData[FriendshipFields.SENDER_NAME] || '';
        friendUsername = friendshipData[FriendshipFields.SENDER_USERNAME] || '';
        friendAvatar = friendshipData[FriendshipFields.SENDER_AVATAR] || '';
      } else {
        friendName = friendshipData[FriendshipFields.RECEIVER_NAME] || '';
        friendUsername =
          friendshipData[FriendshipFields.RECEIVER_USERNAME] || '';
        friendAvatar = friendshipData[FriendshipFields.RECEIVER_AVATAR] || '';
      }

      const friend: Friend = {
        user_id: senderId,
        username: friendUsername,
        name: friendName,
        avatar: friendAvatar,
      };

      // Create analytics event
      const event: InviteEventParams = {
        friend_count: currentUserFriendCount,
        invitation_count: currentUserInvitationCount,
      };

      return {
        data: friend,
        status: 200,
        analytics: {
          event: EventName.INVITE_ACCEPTED,
          userId: currentUserId,
          params: event,
        },
      };
    }
  }

  // Create the friendship document using profile data directly
  const currentTime = Timestamp.now();
  const friendshipData = {
    [FriendshipFields.SENDER_ID]: senderId,
    [FriendshipFields.SENDER_NAME]: senderProfile[ProfileFields.NAME] || '',
    [FriendshipFields.SENDER_USERNAME]:
      senderProfile[ProfileFields.USERNAME] || '',
    [FriendshipFields.SENDER_AVATAR]: senderProfile[ProfileFields.AVATAR] || '',
    [FriendshipFields.RECEIVER_ID]: currentUserId,
    [FriendshipFields.RECEIVER_NAME]:
      currentUserProfile[ProfileFields.NAME] || '',
    [FriendshipFields.RECEIVER_USERNAME]:
      currentUserProfile[ProfileFields.USERNAME] || '',
    [FriendshipFields.RECEIVER_AVATAR]:
      currentUserProfile[ProfileFields.AVATAR] || '',
    [FriendshipFields.STATUS]: Status.ACCEPTED,
    [FriendshipFields.CREATED_AT]: currentTime,
    [FriendshipFields.UPDATED_AT]: currentTime,
    [FriendshipFields.MEMBERS]: [senderId, currentUserId],
  };

  // Add operations to batch
  batch.set(friendshipRef, friendshipData);
  batch.delete(invitationRef);

  const senderName =
    senderProfile[ProfileFields.NAME] ||
    senderProfile[ProfileFields.USERNAME] ||
    'Friend';
  const currentUserName =
    currentUserProfile[ProfileFields.NAME] ||
    currentUserProfile[ProfileFields.USERNAME] ||
    'Friend';

  // Create summary for current user about sender
  const summaryIdForCurrentUser = createSummaryId(currentUserId, senderId);
  const summaryRefForCurrentUser = db
    .collection(Collections.USER_SUMMARIES)
    .doc(summaryIdForCurrentUser);
  const summaryDataForCurrentUser = {
    [UserSummaryFields.CREATOR_ID]: senderId,
    [UserSummaryFields.TARGET_ID]: currentUserId,
    [UserSummaryFields.SUMMARY]: FriendPlaceholderTemplates.SUMMARY.replace(
      '<FRIEND_NAME>',
      senderName,
    ),
    [UserSummaryFields.SUGGESTIONS]:
      FriendPlaceholderTemplates.SUGGESTIONS.replace(
        '<FRIEND_NAME>',
        senderName,
      ),
    [UserSummaryFields.LAST_UPDATE_ID]: '',
    [UserSummaryFields.UPDATE_COUNT]: 0,
    [UserSummaryFields.CREATED_AT]: currentTime,
    [UserSummaryFields.UPDATED_AT]: currentTime,
  };
  batch.set(summaryRefForCurrentUser, summaryDataForCurrentUser);

  // Create summary for sender about current user
  const summaryIdForSender = createSummaryId(senderId, currentUserId);
  const summaryRefForSender = db
    .collection(Collections.USER_SUMMARIES)
    .doc(summaryIdForSender);
  const summaryDataForSender = {
    [UserSummaryFields.CREATOR_ID]: currentUserId,
    [UserSummaryFields.TARGET_ID]: senderId,
    [UserSummaryFields.SUMMARY]: FriendPlaceholderTemplates.SUMMARY.replace(
      '<FRIEND_NAME>',
      currentUserName,
    ),
    [UserSummaryFields.SUGGESTIONS]:
      FriendPlaceholderTemplates.SUGGESTIONS.replace(
        '<FRIEND_NAME>',
        currentUserName,
      ),
    [UserSummaryFields.LAST_UPDATE_ID]: '',
    [UserSummaryFields.UPDATE_COUNT]: 0,
    [UserSummaryFields.CREATED_AT]: currentTime,
    [UserSummaryFields.UPDATED_AT]: currentTime,
  };
  batch.set(summaryRefForSender, summaryDataForSender);

  // Commit the batch
  await batch.commit();

  logger.info(
    `User ${currentUserId} accepted invitation ${invitationId} from ${senderId}`,
  );

  // Return the friend object using sender's profile data
  const friend: Friend = {
    user_id: senderId,
    username: senderProfile[ProfileFields.USERNAME] || '',
    name: senderProfile[ProfileFields.NAME] || '',
    avatar: senderProfile[ProfileFields.AVATAR] || '',
  };

  // Create analytics event
  const event: InviteEventParams = {
    friend_count: currentUserFriendCount + 1, // Add 1 for the new friend
    invitation_count: currentUserInvitationCount,
  };

  return {
    data: friend,
    status: 200,
    analytics: {
      event: EventName.INVITE_ACCEPTED,
      userId: currentUserId,
      params: event,
    },
  };
};
